File: src/csharp/Foundgine.Core/Semantic/Query/SemanticValueValidator.cs
namespace Foundgine.Core.Semantic.Query;

internal static class SemanticValueValidator
{
    public static void Validate(object? value, SemanticField field, string operation)
    {
        if (value is null) return;

        if (value is System.Collections.IEnumerable values && value is not string &&
            operation.Equals("IN", StringComparison.OrdinalIgnoreCase))
        {
            foreach (var item in values)
                ValidateSingle(item, field, operation);
            return;
        }

        ValidateSingle(value, field, operation);
    }

    private static void ValidateSingle(object? value, SemanticField field, string operation)
    {
        if (value is null) return;

        var actual = SemanticType.FromClrType(value.GetType());
        var expected = field.EffectiveSemanticType;
        if (AreCompatible(expected, actual)) return;

        throw new InvalidOperationException(
            $"Semantic {operation} value for field '{field.Name}' has type '{actual}', but the field requires '{expected}'.");
    }

    private static bool AreCompatible(SemanticType expected, SemanticType actual)
    {
        if (expected == actual) return true;
        if (expected is SemanticType.Scalar { Kind: var e } && actual is SemanticType.Scalar { Kind: var a })
        {
            if (e is SemanticScalarKind.Int32 or SemanticScalarKind.Int64 &&
                a is SemanticScalarKind.Int32 or SemanticScalarKind.Int64)
                return true;
            if (e == SemanticScalarKind.Decimal &&
                a is SemanticScalarKind.Int32 or SemanticScalarKind.Int64 or SemanticScalarKind.Decimal)
                return true;
        }

        return false;
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/query/SemanticValueValidator.java
package com.foundgine.core.semantic.query;

import com.foundgine.core.semantic.*;
import java.util.*;

public final class SemanticValueValidator {
	private SemanticValueValidator() {
	}

	public static void validate(Object value, SemanticField field, String operation) {
		if (value == null)
			return;
		if (value instanceof Iterable<?> it && operation.equalsIgnoreCase("IN")) {
			for (var x : it)
				validateSingle(x, field, operation);
			return;
		}
		validateSingle(value, field, operation);
	}

	private static void validateSingle(Object value, SemanticField field, String operation) {
		if (value == null)
			return;
		var actual = SemanticType.fromClass(value.getClass());
		var expected = field.effectiveSemanticType();
		if (compatible(expected, actual))
			return;
		throw new IllegalStateException("Semantic " + operation + " value for field '" + field.name() + "' has type '"
				+ actual + "', but the field requires '" + expected + "'.");
	}

	private static boolean compatible(SemanticType e, SemanticType a) {
		if (e.equals(a))
			return true;
		if (e instanceof SemanticType.Scalar es && a instanceof SemanticType.Scalar as) {
			if ((es.kind() == SemanticScalarKind.INT32 || es.kind() == SemanticScalarKind.INT64)
					&& (as.kind() == SemanticScalarKind.INT32 || as.kind() == SemanticScalarKind.INT64))
				return true;
			if (es.kind() == SemanticScalarKind.DECIMAL && (as.kind() == SemanticScalarKind.INT32
					|| as.kind() == SemanticScalarKind.INT64 || as.kind() == SemanticScalarKind.DECIMAL))
				return true;
		}
		return false;
	}
}
